samples/bluetooth/mesh/sensor_server/src/model_handler.c

```diff
@@ -0,0 +1,225 @@
+#include <dk_buttons_and_leds.h>
+#include "model_handler.h"
+
+#define TEMP_RANGE_COLUMNS_THRESHOLD 32
+
+/* The columns (temperature ranges) for relative
+ * runtime in a chip temperature
+ */
+static const struct bt_mesh_sensor_column columns[] = {
+	{ { 0 }, { TEMP_RANGE_COLUMNS_THRESHOLD } },
+	{ { TEMP_RANGE_COLUMNS_THRESHOLD }, { 100 } },
+};
+
+static struct device *dev;
+static u32_t tot_temp_samps;
+static u32_t col_samps[ARRAY_SIZE(columns)];
+static struct sensor_value relative_runtime[ARRAY_SIZE(columns)];
+
+static s32_t prev_pres;
+
+static int chip_temp_get(struct bt_mesh_sensor *sensor,
+			 struct bt_mesh_msg_ctx *ctx, struct sensor_value *rsp)
+{
+	int err;
+
+	sensor_sample_fetch(dev);
+
+	err = sensor_channel_get(dev, SENSOR_CHAN_DIE_TEMP, rsp);
+
+	if (rsp->val1 < TEMP_RANGE_COLUMNS_THRESHOLD) {
+		col_samps[0]++;
+	} else {
+		col_samps[1]++;
+	}
+
+	tot_temp_samps++;
+
+	return err;
+}
+
+static struct bt_mesh_sensor chip_temp = {
+	.type = &bt_mesh_sensor_present_dev_op_temp,
+	.get = chip_temp_get,
+};
+
+static int relative_runtime_in_chip_temp_get(
+	struct bt_mesh_sensor *sensor, struct bt_mesh_msg_ctx *ctx,
+	const struct bt_mesh_sensor_column *column, struct sensor_value *value)
+{
+	s32_t index = column - &columns[0];
+
+	if (col_samps[index] > 0) {
+		u8_t percent_steps = (200 * col_samps[index]) / tot_temp_samps;
+
+		relative_runtime[index].val1 = percent_steps / 2;
+		relative_runtime[index].val2 = (percent_steps % 2) * 500000;
+	}
+
+	value[0] = relative_runtime[index];
+	value[1] = column->start;
+	value[2] = column->end;
+
+	return 0;
+}
+
+static struct bt_mesh_sensor rel_chip_temp_runtime = {
+	.type = &bt_mesh_sensor_rel_runtime_in_a_dev_op_temp_range,
+	.series = {
+		columns,
+		ARRAY_SIZE(columns),
+		relative_runtime_in_chip_temp_get,
+		},
+};
+
+static struct bt_mesh_sensor presence_sensor = {
+	.type = &bt_mesh_sensor_presence_detected,
+};
+
+static int time_since_presence_detected_get(struct bt_mesh_sensor *sensor,
+					    struct bt_mesh_msg_ctx *ctx,
+					    struct sensor_value *rsp)
+{
+	if (prev_pres) {
+		rsp->val1 = ((k_uptime_get_32() - prev_pres) + 500) / 1000;
+	} else {
+		rsp->val1 = 0;
+	}
+
+	return 0;
+}
+
+static struct bt_mesh_sensor time_since_presence_detected = {
+	.type = &bt_mesh_sensor_time_since_presence_detected,
+	.get = time_since_presence_detected_get,
+};
+
+static struct bt_mesh_sensor *const sensors[] = {
+	&chip_temp,
+	&rel_chip_temp_runtime,
+	&presence_sensor,
+	&time_since_presence_detected,
+};
+
+static struct bt_mesh_sensor_srv sensor_srv =
+	BT_MESH_SENSOR_SRV_INIT(sensors, ARRAY_SIZE(sensors));
+
+static void button_handler_cb(u32_t pressed, u32_t changed)
+{
+	if ((pressed & BIT(0))) {
+		int err;
+
+		/* This sensor value must be boolean -
+		 * .val1 can only be '0' or '1'
+		 */
+		struct sensor_value val = {
+			.val1 = 1,
+		};
+
+		err = bt_mesh_sensor_srv_pub(&sensor_srv, NULL,
+					     &presence_sensor, &val);
```

Comment:
I think you might want to publish again after some delay here. Set a k_delayed_work to fire after 2 seconds or so, that publishes `val.val1 = 0`, to emulate end of presence.

Reply:
Note: Handle this correctly in Client

Reply:
Done